### PR TITLE
Delete operation CodeReady workspace

### PR DESCRIPTION
**Pull request by MaazMS**
Created Delete operation of code ready workspace. 

log for create and delete method  
* create logs 
```bash
2021-10-01T14:28:41.932+0530	INFO	controllers.Workshop	Adding Finalizer for the Workshop	{"workshop": "workshop-infra/cloud-native-workshop"}
INFO[0036] OpenShift Console URL https://console-openshift-console.apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:260"
INFO[0036] Apps Hostname Suffix apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:265"
INFO[0036] Created redis Secret                          source="portal_reconciler.go:51"
INFO[0036] Created redis Persistent Volume Claim         source="portal_reconciler.go:58"
INFO[0036] Created redis Deployment                      source="portal_reconciler.go:66"
INFO[0037] Created redis Service                         source="portal_reconciler.go:87"
INFO[0037] Created portal Deployment                     source="portal_reconciler.go:109"
INFO[0037] Created portal Service                        source="portal_reconciler.go:131"
INFO[0037] Created portal Route                          source="portal_reconciler.go:139"
INFO[0038] Created gitea Project                         source="gitea_reconciler.go:73"
INFO[0039] Created gitea-operator Service Account        source="gitea_reconciler.go:89"
INFO[0039] Created gitea-operator Operator               source="gitea_reconciler.go:114"
INFO[0040] Created gitea-server Custom Resource          source="gitea_reconciler.go:122"
INFO[0040] Waiting for deployment gitea-server. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0142] Deployment 'gitea-server' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0142] Creating CodeReadyWorkspace                   source="codeready_reconciler.go:51"
INFO[0142] Creating CodeReadyWorkspace                   source="codeready_reconciler.go:66"
INFO[0143] Created workspaces CodeReadyWorkspace Project  source="codeready_reconciler.go:75"
INFO[0143] Created codeready-workspaces CodeReadyWorkspace OperatorGroup  source="codeready_reconciler.go:83"
INFO[0143] Created codeready-workspaces CodeReadyWorkspace Subscription  source="codeready_reconciler.go:92"
WARN[0143] Waiting for CodeReadyWorkspace Subscription to create InstallPlan for codeready-workspaces  source="codeready_reconciler.go:97"
INFO[0143] OpenShift Console URL https://console-openshift-console.apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:260"
INFO[0143] Apps Hostname Suffix apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:265"
INFO[0146] Waiting for deployment gitea-server. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0146] Deployment 'gitea-server' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0147] Creating CodeReadyWorkspace                   source="codeready_reconciler.go:51"
INFO[0147] Creating CodeReadyWorkspace                   source="codeready_reconciler.go:66"
INFO[0148] install-brdmb InstallPlan in workspaces project Approved  source="installplan.go:37"
INFO[0148] Waiting for deployment codeready-operator. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0172] Deployment 'codeready-operator' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0173] Waiting for deployment codeready. Default timeout: 420 seconds  source="k8s_helpers.go:68"
ERRO[0593] Failed to verify a successful codeready deployment  source="k8s_helpers.go:98"
ERRO[0593] Failed to find pod to exec into. List of pods: []  source="k8s_helpers.go:152"
ERRO[0593] Pod error log: resource name may not be empty  source="k8s_helpers.go:131"
ERRO[0593] Command to get deployment logs: kubectl logs deployment/codeready -n=workspaces  source="k8s_helpers.go:108"
ERRO[0593] Get k8s events: kubectl get events --field-selector involvedObject.name=$(kubectl get pods -l=component=codeready -n=workspaces -o=jsonpath='{.items[0].metadata.name}') -n=workspaces  source="k8s_helpers.go:109"
INFO[0593] OpenShift Console URL https://console-openshift-console.apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:260"
INFO[0593] Apps Hostname Suffix apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:265"
INFO[0596] Waiting for deployment gitea-server. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0596] Deployment 'gitea-server' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0597] Creating CodeReadyWorkspace                   source="codeready_reconciler.go:51"
INFO[0597] Creating CodeReadyWorkspace                   source="codeready_reconciler.go:66"
INFO[0597] Waiting for deployment codeready-operator. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0598] Deployment 'codeready-operator' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0598] Created codereadyworkspaces CodeReadyWorkspace Custom Resource  source="codeready_reconciler.go:110"
INFO[0598] Waiting for deployment codeready. Default timeout: 420 seconds  source="k8s_helpers.go:68"
	INFO[0911] Deployment 'codeready' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0913] Created che CodeReadyWorkspace Cluster Role   source="codeready_reconciler.go:137"
INFO[0913] Created che CodeReadyWorkspace Cluster Role Binding  source="codeready_reconciler.go:145"
INFO[0914] Created user1 in CodeReady Workspaces         source="codeready_reconciler.go:282"
2021-10-01T14:43:27.471+0530	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "workshop.stakater.com", "reconcilerKind": "Workshop", "controller": "workshop", "name": "cloud-native-workshop", "namespace": "workshop-infra"}
INFO[0921] OpenShift Console URL https://console-openshift-console.apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:260"
INFO[0921] Apps Hostname Suffix apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:265"
INFO[0924] Waiting for deployment gitea-server. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0924] Deployment 'gitea-server' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0924] Creating CodeReadyWorkspace                   source="codeready_reconciler.go:51"
INFO[0924] Creating CodeReadyWorkspace                   source="codeready_reconciler.go:66"
INFO[0925] Waiting for deployment codeready-operator. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0925] Deployment 'codeready-operator' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0926] Waiting for deployment codeready. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0926] Deployment 'codeready' successfully scaled to 1  source="k8s_helpers.go:81"
2021-10-01T14:43:35.147+0530	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "workshop.stakater.com", "reconcilerKind": "Workshop", "controller": "workshop", "name": "cloud-native-workshop", "namespace": "workshop-infra"}
```
* delete logs  
```bash
2021-10-01T14:44:13.826+0530	INFO	controllers.Workshop	Successfully finalized workshop	{"workshop": "workshop-infra/cloud-native-workshop"}
INFO[0967] OpenShift Console URL https://console-openshift-console.apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:260"
INFO[0967] Apps Hostname Suffix apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:265"
2021-10-01T14:44:13.826+0530	INFO	controllers.Workshop	Deleting workshopcloud-native-workshop	{"workshop": "workshop-infra/cloud-native-workshop"}
INFO[0967] Deleting  CodeReadyWorkspace                  source="codeready_reconciler.go:601"
INFO[0967] Deleted che CodeReadyWorkspace Cluster RoleBinding   source="codeready_reconciler.go:613"
INFO[0968] Deleted che CodeReadyWorkspace Cluster Role   source="codeready_reconciler.go:619"
INFO[0968] Deleted codereadyworkspaces codeReadyWorkspaces CustomResource  source="codeready_reconciler.go:626"
INFO[0968] Deleted codeready-workspaces codeReadyWorkspaces Subscription  source="codeready_reconciler.go:634"
INFO[0968] Deleted codeready-workspaces codeReadyWorkspaces OperatorGroup  source="codeready_reconciler.go:641"
INFO[0969] Deleted workspaces codeReadyWorkspaces namespace  source="codeready_reconciler.go:648"
INFO[0969] Deleted  CodeReadyWorkspace Success           source="codeready_reconciler.go:649"
INFO[0969] Deleting gitea                                source="gitea_reconciler.go:202"
INFO[0969] Deleted gitea-server gitea Custom Resource    source="gitea_reconciler.go:212"
INFO[0969] Deleted gitea-operator gitea Operator         source="gitea_reconciler.go:219"
INFO[0970] Deleted gitea-operator gitea Cluster  Role Binding  source="gitea_reconciler.go:226"
INFO[0970] Deleted gitea-operator gitea Cluster Role     source="gitea_reconciler.go:233"
INFO[0970] Deleted gitea-operator gitea Service Account  source="gitea_reconciler.go:240"
INFO[0970] Deleted giteas.gpte.opentlc.com gitea Custom Resource Definition  source="gitea_reconciler.go:247"
INFO[0970] Deleted gitea gitea Project                   source="gitea_reconciler.go:254"
INFO[0970] Gitea deleted succesfully                     source="gitea_reconciler.go:255"
INFO[0970] Deleting nexus                                source="nexus_reconciler.go:128"
I1001 14:44:18.424540  100134 request.go:621] Throttling request took 1.047745043s, request: GET:https://api.workshop.6hf8sw9d.kubeapp.cloud:6443/apis/k8s.cni.cncf.io/v1?timeout=32s
INFO[0974] Finalizer removed for workshopcloud-native-workshop  source="workshop_controller.go:107"
2021-10-01T14:44:20.882+0530	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "workshop.stakater.com", "reconcilerKind": "Workshop", "controller": "workshop", "name": "cloud-native-workshop", "namespace": "workshop-infra"}
INFO[0974] Workshop resource not found. Ignoring since object must be deleted  source="workshop_controller.go:81"
2021-10-01T14:44:20.893+0530	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "workshop.stakater.com", "reconcilerKind": "Workshop", "controller": "workshop", "name": "cloud-native-workshop", "namespace": "workshop-infra"}
```

### CODE REVIEW

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-29-1bbea08f`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-29-9e77e7c7`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-29-24d4a2b1`

File: controllers/codeready_reconciler.go [hunk 1]
```diff
@@ -26,14 +26,33 @@ import (
...
+}
+
+const (
+	CODEREADYNAMESPACENAME          = "workspaces"
+	CODEREADYOPERATORGROUPNAME      = "codeready-workspaces"
+	CODEREADYREDHATSUBSCRIPTIONNAME = "codeready-workspaces"
+	CODEREADYPACKAGENAME            = "codeready-workspaces"
+	CODEREADYDEPLOYMENTNAME         = "codeready-operator"
+	CODEREADYCUSTOMRESOURCENAME     = "codereadyworkspaces"
+	CODEREADYDEPLOYMENTSTATUSNAME   = "codeready"
```
Comment (rasheedamir) [inline controllers/codeready_reconciler.go]: @MaazMS change this to CODEREADYDEPLOYMENTNAME

File: controllers/codeready_reconciler.go [hunk 2]
```diff
@@ -581,86 +596,57 @@ func initWorkspace(workshop *workshopv1.Workshop, username string,
...
 	}
+	log.Infof("Deleted %s codeReadyWorkspaces OperatorGroup", codeReadyWorkspacesOperatorGroup.Name)
 
+	codeReadyWorkspacesNamespace := kubernetes.NewNamespace(workshop, r.Scheme, CODEREADYNAMESPACENAME)
+	// Delete Project
+	if err := r.Delete(context.TODO(), codeReadyWorkspacesNamespace); err != nil {
+		return reconcile.Result{}, err
+	}
+	log.Infof("Deleted %s codeReadyWorkspaces namespace", codeReadyWorkspacesNamespace.Name)
+	log.Infoln("Deleted  CodeReadyWorkspace Success")
```
Comment (rasheedamir) [inline controllers/codeready_reconciler.go]: successfully

File: controllers/codeready_reconciler.go (hunk 1 above)
Comment (MaazMS) [reply controllers/codeready_reconciler.go]: @rasheedamir should i  change CODEREADYDEPLOYMENTNAME to CODEREADYDEPLOYMENTSTATUS

File: controllers/codeready_reconciler.go (hunk 2 above)
Comment (MaazMS) [reply controllers/codeready_reconciler.go]: yes i will change Success to successfully

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-29-60dd79d2`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-29-609efd2d`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-29-578cf991`

File: controllers/codeready_reconciler.go [hunk 3]
```diff
@@ -26,14 +26,39 @@ import (
 	"sigs.k8s.io/yaml"
 )
 
+var CodeReadyLabels = map[string]string{
```
Comment (tanalam2411) [inline controllers/codeready_reconciler.go]: If `CodeReadyLabels` is not getting used outside the package, please make it `codeReadyLabels`

File: controllers/codeready_reconciler.go [hunk 4]
```diff
@@ -132,17 +153,19 @@ func (r *WorkshopReconciler) addCodeReadyWorkspace(workshop *workshopv1.Workshop
 
 		for id := 1; id <= users; id++ {
 			username := fmt.Sprintf("user%d", id)
-
-			if result, err := createUser(workshop, username, "codeready", codeReadyWorkspacesNamespace.Name, appsHostnameSuffix, masterAccessToken); err != nil {
+			log.Infoln("user name ", username)
+			log.Infoln("namespace name", CODEREADY_NAMESPACE_NAME)
+			log.Infoln(username + CODEREADY_NAMESPACE_NAME)
```
Comment (tanalam2411) [inline controllers/codeready_reconciler.go]: Remove all these 3 logs

**Review by tanalam2411 — CHANGES_REQUESTED** [review]
@MaazMS Please make the required changes

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-29-d5dcd64f`

File: controllers/codeready_reconciler.go [hunk 5]
```diff
@@ -581,86 +594,73 @@ func initWorkspace(workshop *workshopv1.Workshop, username string,
...
+	if !workshop.Spec.Infrastructure.CodeReadyWorkspace.OpenshiftOAuth {
+
+		for id := 1; id <= users; id++ {
+			username := fmt.Sprintf("user%d", id)
+
+			userWorkspacesNamespaceName := username + "-" + "workspace"
+			userWorkspacesNamespace := kubernetes.NewNamespace(workshop, r.Scheme, userWorkspacesNamespaceName)
+			// Delete Project
+			if err := r.Delete(context.TODO(), userWorkspacesNamespace); err != nil {
+				log.Errorf("Failed to Deleted %s Namespace", userWorkspacesNamespace.Name)
```
Comment (tanalam2411) [inline controllers/codeready_reconciler.go]: It should be `Failed to Delete` not `Failed to Deleted`

**Review by tanalam2411 — CHANGES_REQUESTED** [review]
@MaazMS, Please fix the log message

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-29-2882606f`

**Review by tanalam2411 — APPROVED** [review]
LGTM

**Pull request merged**